lib/lib-upload/src/main/java/org/sagebionetworks/upload/multipart/GoogleCloudStorageMultipartUploadDAOImpl.java

```diff
@@ -0,0 +1,122 @@
+	@Override
+	public void validateAndAddPart(AddPartRequest request) {
+		validatePartMd5(request);
+		addPart(request.getUploadId(), request.getBucket(), request.getKey(), request.getPartNumber(), request.getPartNumber(), request.getTotalNumberOfParts());
+	}
+
+	void validatePartMd5(AddPartRequest request) {
+		Blob uploadedPart = googleCloudStorageClient.getObject(request.getBucket(), request.getPartKey());
+		if (uploadedPart == null) {
+			throw new IllegalArgumentException("The uploaded part could not be found");
+		}
+		if (!Hex.encodeHexString(Base64.decodeBase64(uploadedPart.getMd5())).equals(request.getPartMD5Hex())) {
+			throw new IllegalArgumentException("The provided MD5 does not match the MD5 of the uploaded part.  Please re-upload the part.");
+		}
+		// The part was uploaded successfully
+	}
+
+	void addPart(String uploadId, String bucket, String key, Long lowerBound, Long upperBound, Long totalNumberOfParts) {
+		multipartUploadComposerDAO.addPartToUpload(uploadId, lowerBound, upperBound);
+		// If lowerBound is 1 and upperBound == totalNumber, then we have the entire file. Otherwise, try to stitch
+		if (lowerBound != 1 || !upperBound.equals(totalNumberOfParts)) {
+			PartRange expectedStitchPartRange =
+					MultipartUploadUtils.getRangeOfPotentialStitchTargets(lowerBound, upperBound, totalNumberOfParts);
+
+			// Get the parts that have already been uploaded
+			List<DBOMultipartUploadComposerPartState> uploadedParts = multipartUploadComposerDAO
+					.getAddedPartRanges(Long.valueOf(uploadId),
+							expectedStitchPartRange.getLowerBound(),
+							expectedStitchPartRange.getUpperBound());
+
+			// We must make sure that the sum of each part size is equivalent to the size of a new merged part
+			long sumOfUploadedPartSizesInRange = 0;
+			for (DBOMultipartUploadComposerPartState part : uploadedParts) {
+				sumOfUploadedPartSizesInRange += part.getSizeOfPart();
+			}
+
+			if (expectedStitchPartRange.getSize() == sumOfUploadedPartSizesInRange // Make sure that amount of data in the new part is equivalent to the amount of data we have
+					&& expectedStitchPartRange.getNumberOfParts() == uploadedParts.size()){	// We must also make sure that we get exactly the number of parts that we expect
+
+				// We can now stitch the parts
+				// Get the key names for the old parts
+				List<String> partKeys = new ArrayList<>();
+				for (DBOMultipartUploadComposerPartState part : uploadedParts) {
+					partKeys.add(MultipartUploadUtils.createPartKeyFromRange(key, part.getPartRangeLowerBound(), part.getPartRangeUpperBound()));
+				}
+
+				// Create a new part
+				googleCloudStorageClient.composeObjects(bucket,
+						MultipartUploadUtils.createPartKeyFromRange(key, expectedStitchPartRange.getLowerBound(), expectedStitchPartRange.getUpperBound()),
+						partKeys);
+
+				// Delete the old parts from Google Cloud and the database.
+				for (String part : partKeys) {
+					googleCloudStorageClient.deleteObject(bucket, part);
```

Comment:
move delete to complete multi-part upload, else error thrown deleting would make this part unrecoverable.